spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/PushDown.scala

```diff
@@ -0,0 +1,204 @@
+package org.apache.spark.sql.cassandra
+
+import com.datastax.spark.connector.cql.TableDef
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.sources.Filter
+
+/**
+ * Calculate pushdown predicates for a given table and Catalyst [[Expression]]s
+ */
+class PredicatePushDown (predicates: Seq[Expression], tableDef: TableDef)
+  extends PushDown(predicates: Seq[Expression], tableDef: TableDef) {
+
+  def isEqualTo(predicate: Expression): Boolean = predicate.isInstanceOf[EqualTo]
+
+  def isIn(predicate: Expression) : Boolean = predicate.isInstanceOf[In] || predicate.isInstanceOf[InSet]
+
+  def isRangeComparison(predicate: Expression) : Boolean = predicate match {
+    case _: LessThan           => true
+    case _: LessThanOrEqual    => true
+    case _: GreaterThan        => true
+    case _: GreaterThanOrEqual => true
+    case _                     => false
+  }
+
+  def isSingleColumn(predicate: Expression) = predicate.references.size == 1
+
+  /** Returns the only column name referenced in the predicate */
+  def predicateColumnName(predicate: Expression) : String = {
+    require(predicate.references.size == 1, "Given predicate is not a single column predicate: " + predicate)
+    predicate.references.head.name
+  }
+}
+
+/**
+ * Calculate pushdown filters for a given table and source [[Filter]]s
+ */
+class FilterPushDown (filters: Seq[Filter], tableDef: TableDef)
+  extends PushDown(filters: Seq[Filter], tableDef: TableDef) {
+
+  def isEqualTo(filter: Filter): Boolean = filter.isInstanceOf[sources.EqualTo]
+
+  def isIn(filter: Filter) : Boolean = filter.isInstanceOf[sources.In]
+
+  def isRangeComparison(filter: Filter) : Boolean = filter match {
+    case _: sources.LessThan           => true
+    case _: sources.LessThanOrEqual    => true
+    case _: sources.GreaterThan        => true
+    case _: sources.GreaterThanOrEqual => true
+    case _                             => false
+  }
+
+  def isSingleColumn(filter: Filter) = filter.isInstanceOf[sources.In] || filter.isInstanceOf[sources.EqualTo] ||
```

Comment:
If you implemented this function in this way, it could be moved to the super class and used by both implementations (which would be safer for `PredicatePushDown`).